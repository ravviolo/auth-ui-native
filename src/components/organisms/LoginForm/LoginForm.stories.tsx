import { action } from '@storybook/addon-actions';
import { text } from '@storybook/addon-knobs';
import { storiesOf } from '@storybook/react-native';
import { View } from 'react-native';
import { StoryScreen } from 'storybook/StoryScreen';

import { LoginForm } from './LoginForm';

storiesOf('LoginForm', module)
  .addDecorator((getStory) => (
    <StoryScreen>
      <View style={{ flex: 0.3 }}>{getStory()}</View>
    </StoryScreen>
  ))
  .add('Default', () => (
    <LoginForm
      testID={text('testID', 'login-form-test-id')}
      onPressResetPassword={action('onPressResetPassword')}
      onSubmitLoginForm={action('onSubmitLoginForm')}
    />
  ));
