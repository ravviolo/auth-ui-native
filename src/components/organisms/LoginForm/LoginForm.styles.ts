import { StyleSheet } from 'react-native';
import { colors } from 'theme';

export const styles = StyleSheet.create({
  checkboxContainer: {
    justifyContent: 'space-between',
    width: 130,
  },
  container: {
    flex: 1,
    justifyContent: 'space-between',
  },
  secondaryButton: {
    alignSelf: 'flex-end',
    color: colors.textLight,
  },
});
