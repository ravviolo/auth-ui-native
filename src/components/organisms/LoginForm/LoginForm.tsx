import {
  Button,
  Checkbox,
  CheckboxHandle,
  Flex,
  InputLabel,
  TextButton,
  TextInputHandle,
} from 'components/atoms';
import { EmailInput, PasswordInput } from 'components/molecules';
import { useEffect, useRef } from 'react';

import { styles } from './LoginForm.styles';

export interface Props {
  onSubmitLoginForm: (email: string, password: string, rememberPassword: boolean) => void;
  onPressResetPassword: () => void;
}

export const LoginForm = ({ onSubmitLoginForm, onPressResetPassword }: Props) => {
  const emailRef = useRef<TextInputHandle | null>(null);
  const passwordRef = useRef<TextInputHandle | null>(null);
  const rememberPasswordRef = useRef<CheckboxHandle | null>(null);

  useEffect(() => {
    emailRef.current?.setFocus();
  }, []);
  // todo: console.log('login render');

  const onSubmit = () => {
    const email = emailRef.current?.getValue();
    const password = passwordRef.current?.getValue();
    const rememberPassword = !!rememberPasswordRef.current?.getValue();

    if (email && password) {
      onSubmitLoginForm(email, password, rememberPassword);
      emailRef.current?.clear();
      passwordRef.current?.clear();
      rememberPasswordRef.current?.clear();
    }
  };
  return (
    <Flex direction="column" style={styles.container}>
      <EmailInput ref={emailRef} testID="email-input-login-form-test-id" />
      <PasswordInput ref={passwordRef} testID="password-input-login-form-test-id" />
      <Flex
        direction="row"
        style={styles.checkboxContainer}
        testID="checkbox-container-login-form-test-id"
      >
        <Checkbox ref={rememberPasswordRef} testID="remember-password-checkbox-test-id" />
        <InputLabel label="Remember me?" testID="input-label-remember-password-checkbox-test-id" />
      </Flex>
      <Button testID="btn-submit-login-test-id" title="Login" onPress={onSubmit} />
      <TextButton
        style={styles.secondaryButton}
        testID="btn-reset-password-test-id"
        title="Forgot Password?"
        onPress={onPressResetPassword}
      />
    </Flex>
  );
};
