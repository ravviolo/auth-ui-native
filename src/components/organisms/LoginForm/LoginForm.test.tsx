import { fireEvent, render, screen, within } from '@testing-library/react-native';
import testRenderer from 'react-test-renderer';

import { LoginForm } from './LoginForm';

import type { Props } from './LoginForm';

describe('LoginForm', () => {
  const mockOnSubmitLoginForm = jest.fn<Props['onSubmitLoginForm'], [string, string, boolean]>();
  const mockOnPressResetPassword = jest.fn<Props['onPressResetPassword'], []>();
  const testEmail = 'test@test.com';
  const testPassword = 'testpassword';

  const props: Props = {
    onPressResetPassword: mockOnPressResetPassword,
    onSubmitLoginForm: mockOnSubmitLoginForm,
  };

  it('should render email and password inputs', () => {
    render(<LoginForm {...props} />);

    const emailInput = screen.getByTestId('email-input-login-form-test-id');
    const passwordInput = screen.getByTestId('password-input-login-form-test-id');

    expect(emailInput).not.toBeNull();
    expect(passwordInput).not.toBeNull();
  });

  it("should render checkbox with 'Remember me?' label", () => {
    render(<LoginForm {...props} />);

    const checkbox = screen.getByTestId('remember-password-checkbox-test-id');
    const checkboxLabel = screen.getByTestId('input-label-remember-password-checkbox-test-id');

    expect(checkboxLabel).toHaveTextContent(/Remember me?/);
    expect(checkbox).not.toBeNull();
  });

  it('should render initially unchecked checkbox', () => {
    render(<LoginForm {...props} />);

    const checkbox = screen.getByTestId('remember-password-checkbox-test-id');

    expect(checkbox.props.accessibilityState).toHaveProperty('checked', false);
  });

  it("should render submit button with 'Login' text", () => {
    render(<LoginForm {...props} />);

    const submitButton = screen.getByTestId('btn-submit-login-test-id');

    expect(submitButton).toHaveTextContent('Login');
  });

  it('should submit form with non-empty input fields and checkbox unchecked - run submit handler once', () => {
    render(<LoginForm {...props} />);

    const submitButton = screen.getByTestId('btn-submit-login-test-id');
    const emailInputField = within(
      screen.getByTestId('email-input-login-form-test-id')
    ).getByTestId('input-field-email-test-id');
    const passwordInputField = within(
      screen.getByTestId('password-input-login-form-test-id')
    ).getByTestId('input-field-password-test-id');

    fireEvent.changeText(emailInputField, testEmail);
    fireEvent.changeText(passwordInputField, testPassword);
    fireEvent.press(submitButton);

    expect(mockOnSubmitLoginForm).toHaveBeenCalledTimes(1);
    expect(mockOnSubmitLoginForm).toHaveBeenCalledWith(testEmail, testPassword, false);
  });

  it('should submit form with non-empty input fields and checkbox checked - run submit handler once', () => {
    render(<LoginForm {...props} />);

    const submitButton = screen.getByTestId('btn-submit-login-test-id');
    const emailInputField = within(
      screen.getByTestId('email-input-login-form-test-id')
    ).getByTestId('input-field-email-test-id');
    const passwordInputField = within(
      screen.getByTestId('password-input-login-form-test-id')
    ).getByTestId('input-field-password-test-id');
    const checkbox = screen.getByTestId('remember-password-checkbox-test-id');

    fireEvent.changeText(emailInputField, testEmail);
    fireEvent.changeText(passwordInputField, testPassword);
    fireEvent.press(checkbox);
    fireEvent.press(submitButton);

    expect(mockOnSubmitLoginForm).toHaveBeenCalledTimes(1);
    expect(mockOnSubmitLoginForm).toHaveBeenCalledWith(testEmail, testPassword, true);
  });

  it("shouldn't submit form with one or more empty input fields", () => {
    render(<LoginForm {...props} />);

    const submitButton = screen.getByTestId('btn-submit-login-test-id');

    fireEvent.press(submitButton);

    expect(mockOnSubmitLoginForm).not.toHaveBeenCalled();

    const emailInputField = within(
      screen.getByTestId('email-input-login-form-test-id')
    ).getByTestId('input-field-email-test-id');

    fireEvent.changeText(emailInputField, testEmail);
    fireEvent.press(submitButton);

    expect(mockOnSubmitLoginForm).not.toHaveBeenCalled();
  });

  it("should render reset password button with 'Forgot Password?' text", () => {
    render(<LoginForm {...props} />);

    const resetPasswordButton = screen.getByTestId('btn-reset-password-test-id');

    expect(resetPasswordButton).toHaveTextContent('Forgot Password?');
  });

  it('reset password button should handle press events - run handler once', () => {
    render(<LoginForm {...props} />);

    const resetPasswordButton = screen.getByTestId('btn-reset-password-test-id');

    fireEvent.press(resetPasswordButton);

    expect(mockOnPressResetPassword).toHaveBeenCalledTimes(1);
  });

  it('should match snapshot', () => {
    const tree = testRenderer.create(<LoginForm {...props} />).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
