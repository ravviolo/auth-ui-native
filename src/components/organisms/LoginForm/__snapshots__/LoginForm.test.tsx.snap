// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginForm should match snapshot 1`] = `
<View
  style={
    Array [
      Object {
        "flexDirection": "column",
      },
      Object {
        "flex": 1,
        "justifyContent": "space-between",
      },
    ]
  }
>
  <View
    testID="email-input-login-form-test-id"
  >
    <Text
      style={
        Object {
          "color": "#3D3D3D",
          "marginBottom": 3,
        }
      }
      testID="input-label-email-test-id"
    >
      Email
    </Text>
    <TextInput
      onBlur={[Function]}
      onChangeText={[Function]}
      onFocus={[Function]}
      secureTextEntry={false}
      style={
        Array [
          Object {
            "borderColor": "#9E9E9E",
            "borderRadius": 8,
            "borderWidth": 1,
            "height": 40,
            "padding": 10,
          },
          Object {
            "borderColor": "#9E9E9E",
          },
        ]
      }
      testID="input-field-email-test-id"
      value=""
    />
  </View>
  <View
    testID="password-input-login-form-test-id"
  >
    <Text
      style={
        Object {
          "color": "#3D3D3D",
          "marginBottom": 3,
        }
      }
      testID="input-label-password-test-id"
    >
      Password
    </Text>
    <TextInput
      onBlur={[Function]}
      onChangeText={[Function]}
      onFocus={[Function]}
      secureTextEntry={true}
      style={
        Array [
          Object {
            "borderColor": "#9E9E9E",
            "borderRadius": 8,
            "borderWidth": 1,
            "height": 40,
            "padding": 10,
          },
          Object {
            "borderColor": "#9E9E9E",
          },
        ]
      }
      testID="input-field-password-test-id"
      value=""
    />
  </View>
  <View
    style={
      Array [
        Object {
          "flexDirection": "row",
        },
        Object {
          "justifyContent": "space-between",
          "width": 130,
        },
      ]
    }
    testID="checkbox-container-login-form-test-id"
  >
    <View
      accessibilityRole="checkbox"
      accessibilityState={
        Object {
          "checked": false,
          "disabled": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "borderColor": "#657786",
            "borderRadius": 2,
            "borderWidth": 2,
            "height": 20,
            "width": 20,
          },
          Object {
            "borderRadius": 4,
            "borderWidth": 1,
          },
          false,
          Object {
            "backgroundColor": undefined,
            "borderColor": "#9E9E9E",
          },
          undefined,
          false,
        ]
      }
      testID="remember-password-checkbox-test-id"
    />
    <Text
      style={
        Object {
          "color": "#3D3D3D",
          "marginBottom": 3,
        }
      }
      testID="input-label-remember-password-checkbox-test-id"
    >
      Remember me?
    </Text>
  </View>
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#FF4684",
        "borderRadius": 10,
        "elevation": 5,
        "padding": 15,
      }
    }
    testID="btn-submit-login-test-id"
  >
    <Text
      style={
        Object {
          "color": "#EDEDED",
          "letterSpacing": 1,
          "textTransform": "uppercase",
        }
      }
    >
      Login
    </Text>
  </View>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    hitSlop={20}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={false}
    testID="btn-reset-password-test-id"
  >
    <Text
      style={
        Object {
          "alignSelf": "flex-end",
          "color": "#9E9E9E",
        }
      }
    >
      Forgot Password?
    </Text>
  </View>
</View>
`;
